diff --git a/library.json b/library.json
index b7f0fd3..e34f1cc 100644
--- a/library.json
+++ b/library.json
@@ -18,12 +18,10 @@
   "dependencies": [
     {
       "name": "ESPAsyncTCP",
-      "version": "1.2.0",
       "platforms": "espressif8266"
     },
     {
       "name": "AsyncTCP",
-      "version": "^1.0.0",
       "platforms": "espressif32"
     }
   ]
diff --git a/src/AsyncMqttClient.cpp b/src/AsyncMqttClient.cpp
index 8fa959d..392c172 100644
--- a/src/AsyncMqttClient.cpp
+++ b/src/AsyncMqttClient.cpp
@@ -2,6 +2,7 @@
 
 AsyncMqttClient::AsyncMqttClient()
 : _connected(false)
+, _lockMutiConnections(false)
 , _connectPacketNotEnoughSpace(false)
 , _disconnectOnPoll(false)
 , _tlsBadFingerprint(false)
@@ -43,7 +44,6 @@ AsyncMqttClient::AsyncMqttClient()
   sprintf(_generatedClientId, "esp8266-%06x", ESP.getChipId());
 #endif
   _clientId = _generatedClientId;
-
   setMaxTopicLength(128);
 }
 
@@ -176,7 +176,7 @@ void AsyncMqttClient::_clear() {
 /* TCP */
 void AsyncMqttClient::_onConnect(AsyncClient* client) {
   (void)client;
-
+  _lockMutiConnections = true;
 #if ASYNC_TCP_SSL_ENABLED
   if (_secure && _secureServerFingerprints.size() > 0) {
     SSL* clientSsl = _client.getSSL();
@@ -354,6 +354,7 @@ void AsyncMqttClient::_onConnect(AsyncClient* client) {
 
 void AsyncMqttClient::_onDisconnect(AsyncClient* client) {
   (void)client;
+  _lockMutiConnections = false;
   AsyncMqttClientDisconnectReason reason;
 
   if (_connectPacketNotEnoughSpace) {
@@ -707,7 +708,8 @@ bool AsyncMqttClient::connected() const {
 
 void AsyncMqttClient::connect() {
   if (_connected) return;
-
+  if (_lockMutiConnections) return;
+  _lockMutiConnections = true;
 #if ASYNC_TCP_SSL_ENABLED
   if (_useIp) {
     _client.connect(_ip, _port, _secure);
@@ -725,7 +727,8 @@ void AsyncMqttClient::connect() {
 
 void AsyncMqttClient::disconnect(bool force) {
   if (!_connected) return;
-
+  if (!_lockMutiConnections) return;
+  _lockMutiConnections = false;
   if (force) {
     _client.close(true);
   } else {
diff --git a/src/AsyncMqttClient.hpp b/src/AsyncMqttClient.hpp
index 4f41210..be58141 100644
--- a/src/AsyncMqttClient.hpp
+++ b/src/AsyncMqttClient.hpp
@@ -82,8 +82,8 @@ class AsyncMqttClient {
 
  private:
   AsyncClient _client;
-
   bool _connected;
+  bool _lockMutiConnections;
   bool _connectPacketNotEnoughSpace;
   bool _disconnectOnPoll;
   bool _tlsBadFingerprint;
   